File: crates/zipp-vm/tests/inline_dense_store.js
"use strict";
// B264: the inline pinned dense-Array store lane. Every loop is hot enough
// for a MEM region with the array pinned; each case is one way the lane must
// route to the helper or deopt, and the answer must match node exactly.
var out = [];
function log(tag, v) { out.push(tag + ":" + v); }

// 1. hole-fill into `new Array(n)` (licensed fill), then overwrite.
var holes = new Array(4096);
var acc = 0;
for (var r = 0; r < 40; r++) {
  for (var i = 0; i < 4096; i++) holes[i] = (i * 7 + r) | 0;
  for (var i = 0; i < 4096; i++) acc = (acc + holes[i]) | 0;
}
log("holes", acc + "," + holes.length + "," + (2 in holes));

// 2. append at i == len (helper path) mixed with in-range stores.
var grow = [];
for (var r = 0; r < 30; r++) {
  for (var i = 0; i < 2000; i++) {
    if (i < grow.length) grow[i] = grow[i] + 1; else grow[i] = i;
  }
}
var gs = 0;
for (var i = 0; i < grow.length; i++) gs = (gs + grow[i]) | 0;
log("grow", grow.length + "," + gs);

// 3. OLD holder across minors: a retained array receives young objects; the
//    barrier must record them (or the lane must go to the helper).
var retained = new Array(512);
for (var i = 0; i < 512; i++) retained[i] = { v: i };
var junk = null;
for (var r = 0; r < 200; r++) {
  for (var i = 0; i < 512; i++) retained[i] = { v: retained[i].v + 1, r: r };
  var pile = [];
  for (var k = 0; k < 400; k++) pile.push({ k: k, s: "s" + k });
  junk = pile;
}
var rs = 0;
for (var i = 0; i < 512; i++) rs = (rs + retained[i].v + retained[i].r) | 0;
log("retained", rs + "," + junk.length);

// 4. non-writable length + hole fill: the store is silently ignored in sloppy
//    mode but THROWS in strict mode (this file is strict).
var fixed = new Array(64);
Object.defineProperty(fixed, "length", { writable: false });
var threw = 0;
for (var r = 0; r < 300; r++) {
  for (var i = 0; i < 64; i++) {
    try { fixed[i] = i; } catch (e) { threw++; }
  }
}
log("fixed", threw + "," + fixed.length + "," + (0 in fixed));

// 5. non-writable length, PRESENT elements: in-range stores are fine.
var fixed2 = [1, 2, 3, 4, 5, 6, 7, 8];
Object.defineProperty(fixed2, "length", { writable: false });
var f2 = 0;
for (var r = 0; r < 4000; r++) for (var i = 0; i < 8; i++) { fixed2[i] = r + i; f2 = (f2 + fixed2[i]) | 0; }
log("fixed2", f2 + "," + fixed2.length);

// 6. an indexed setter on Array.prototype fires for hole fills, not for
//    present-element stores.
var setterHits = 0;
Object.defineProperty(Array.prototype, "3", {
  set: function (v) { setterHits++; },
  get: function () { return "proto3"; },
  configurable: true
});
var withHoles = new Array(8);
var dense = [0, 0, 0, 0, 0, 0, 0, 0];
for (var r = 0; r < 500; r++) {
  for (var i = 0; i < 8; i++) { withHoles[i] = i; dense[i] = i + r; }
  if (r % 2 === 0) withHoles = new Array(8);
}
log("setter", setterHits + "," + withHoles[3] + "," + dense[3]);
delete Array.prototype[3];

// 7. setPrototypeOf(arr, {5: 'x'}) then hole-fill/read.
var protoObj = { 5: "x" };
var sp = new Array(16);
Object.setPrototypeOf(sp, protoObj);
var spv = "";
for (var r = 0; r < 300; r++) {
  sp = new Array(16);
  Object.setPrototypeOf(sp, protoObj);
  for (var i = 0; i < 16; i++) if (i !== 5) sp[i] = i;
  spv = sp[5] + "," + sp[4];
}
log("setproto", spv + "," + sp.length);

// 8. fractional / negative / string keys on a pinned array.
var keyed = new Array(32);
for (var i = 0; i < 32; i++) keyed[i] = 0;
for (var r = 0; r < 500; r++) {
  for (var i = 0; i < 32; i++) {
    keyed[i] = i;
    keyed[i + 0.5] = "half";
    keyed[-1] = "neg";
    keyed["k" + (i & 3)] = i;
  }
}
log("keys", keyed.length + "," + keyed[1.5] + "," + keyed[-1] + "," + keyed.k3 + "," + Object.keys(keyed).length);

// 9. store of a double, a string, undefined and null into a young array.
var mixed = new Array(1024);
var ms = 0;
for (var r = 0; r < 60; r++) {
  for (var i = 0; i < 1024; i++) {
    mixed[i] = (i & 3) === 0 ? i * 1.5 : (i & 3) === 1 ? "s" : (i & 3) === 2 ? undefined : null;
  }
  for (var i = 0; i < 1024; i++) ms = (ms + (typeof mixed[i] === "number" ? mixed[i] : mixed[i] === "s" ? 1 : mixed[i] === undefined ? 2 : 3)) | 0;
}
log("mixed", ms);

console.log(out.join("|"));

File: crates/zipp-vm/tests/inline_dense_store.rs
//! B264: the inline pinned dense-Array `SetIndex` lane in MEM regions
//! (`region_mem.rs`, `ZIPP_NO_INLINE_DENSE_STORE=1` restores the helper-only
//! route). The lane stores directly when the receiver matches its pin
//! snapshot, the key is an in-range integer, the holder is YOUNG (no barrier
//! needed) and the element is present or the hole-fill is licensed by the
//! snapshot flags; every other shape reaches `jit_set_index` exactly as
//! before. The JS file exercises each routing edge -- hole fills, appends, an
//! OLD holder across minors, a non-writable `length`, an indexed setter on
//! `Array.prototype`, a custom prototype, fractional/negative/string keys and
//! non-number values -- and its output is node-oracled (v24.12.0).

const SRC: &str = include_str!("inline_dense_store.js");
const EXPECTED: &str = "holes:-1943535616,4096,true|grow:2000,2057000|retained:335104,400|\
fixed:0,64,true|fixed2:64096000,8|setter:500,proto3,502|setproto:x,4,16|\
keys:32,half,neg,31,69|mixed:11842560";

#[test]
fn inline_dense_store_output_matches_node() {
    let out = zipp_vm::run(SRC).expect("source compiles");
    assert!(out.error.is_none(), "{:?}", out.error);
    assert_eq!(out.output, vec![EXPECTED.to_string()]);
}

/// Exact output in every execution mode, with the latch off, and under the
/// collector torture modes (a barrier hole panics at the minor where it
/// opens under `ZIPP_NURSERY_VERIFY=1`). Each child is a fresh process.
#[test]
fn inline_dense_store_matches_in_every_mode() {
    if std::env::var_os("ZIPP_INLINE_DENSE_STORE_CHILD").is_some() {
        return;
    }
    let exe = std::env::current_exe().expect("test binary path");
    const LATCHES: [&str; 7] = [
        "ZIPP_NO_INLINE_DENSE_STORE",
        "ZIPP_NOJIT",
        "ZIPP_JIT_THRESHOLD",
        "ZIPP_NO_NURSERY",
        "ZIPP_GC_STRESS",
        "ZIPP_NURSERY_VERIFY",
        "ZIPP_GCSTATS",
    ];
    for (mode, env) in [
        ("default", vec![]),
        ("no-inline-dense-store", vec![("ZIPP_NO_INLINE_DENSE_STORE", "1")]),
        ("interpreter", vec![("ZIPP_NOJIT", "1")]),
        ("forced-jit", vec![("ZIPP_JIT_THRESHOLD", "1")]),
        ("no-nursery", vec![("ZIPP_NO_NURSERY", "1")]),
        ("gc-stress", vec![("ZIPP_GC_STRESS", "1")]),
        ("nursery-verify", vec![("ZIPP_NURSERY_VERIFY", "1")]),
        ("nursery-verify+forced-jit", vec![("ZIPP_NURSERY_VERIFY", "1"), ("ZIPP_JIT_THRESHOLD", "1")]),
        ("gcstats-oracle", vec![("ZIPP_GCSTATS", "1")]),
    ] {
        let mut cmd = std::process::Command::new(&exe);
        cmd.args(["inline_dense_store_output_matches_node", "--nocapture"])
            .env("ZIPP_INLINE_DENSE_STORE_CHILD", "1");
        for l in LATCHES {
            cmd.env_remove(l);
        }
        for (key, value) in env {
            cmd.env(key, value);
        }
        let out = cmd.output().expect("spawn mode child");
        assert!(
            out.status.success(),
            "{mode} failed:\n--- stdout ---\n{}\n--- stderr ---\n{}",
            String::from_utf8_lossy(&out.stdout),
            String::from_utf8_lossy(&out.stderr)
        );
    }
}
